services/sso/src/main/java/software.amazon.awssdk.services.sso/internal/SsoProfileCredentialsProviderFactory.java

```diff
@@ -0,0 +1,135 @@
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.services.sso.internal;
+
+import static software.amazon.awssdk.services.sso.internal.SsoCachedTokenFileUtils.generateCachedTokenPath;
+import static software.amazon.awssdk.utils.UserHomeDirectoryUtils.userHomeDirectory;
+
+import java.nio.file.Paths;
+import java.util.function.Supplier;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.annotations.SdkTestInternalApi;
+import software.amazon.awssdk.auth.credentials.AnonymousCredentialsProvider;
+import software.amazon.awssdk.auth.credentials.AwsCredentials;
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
+import software.amazon.awssdk.auth.credentials.SsoCredentialsProviderFactory;
+import software.amazon.awssdk.auth.credentials.internal.SsoCredentialsProfileProperties;
+import software.amazon.awssdk.profiles.Profile;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.sso.SsoClient;
+import software.amazon.awssdk.services.sso.auth.SsoCredentialsProvider;
+import software.amazon.awssdk.services.sso.model.GetRoleCredentialsRequest;
+import software.amazon.awssdk.utils.IoUtils;
+import software.amazon.awssdk.utils.SdkAutoCloseable;
+
+/**
+ * An implementation of {@link SsoCredentialsProviderFactory} that allows users to get SSO role credentials using the startUrl
+ * sprcified in either a {@link Profile} or environment variables.
+ */
+@SdkInternalApi
+public class SsoProfileCredentialsProviderFactory implements SsoCredentialsProviderFactory {
+
+    private static final String TOKEN_DIRECTORY = Paths.get(userHomeDirectory(), ".aws", "sso", "cache").toString();
+
+    private SsoCachedAccessTokenProvider tokenProvider;
+
+    /**
+     * Default method to create the {@link SsoProfileCredentialsProvider} with a {@link SsoCachedAccessTokenProvider}
+     * object created with the start url from {@link Profile} or environment variables and the default token file directory.
+     */
+    public AwsCredentialsProvider create(SsoCredentialsProfileProperties ssoCredentialsProfileProperties) {
+        this.tokenProvider = new SsoCachedAccessTokenProvider(
+            generateCachedTokenPath(ssoCredentialsProfileProperties.ssoStartUrl(), TOKEN_DIRECTORY));
+        return new SsoProfileCredentialsProvider(ssoCredentialsProfileProperties, tokenProvider);
+    }
+
+    /**
+     * Alternative method to create the {@link SsoProfileCredentialsProvider} with a customized
+     * {@link SsoCachedAccessTokenProvider}. This method is only used for testing.
+     */
+    @SdkTestInternalApi
+    public AwsCredentialsProvider create(SsoCredentialsProfileProperties ssoCredentialsProfileProperties,
+                                         SsoCachedAccessTokenProvider tokenProvider) {
```

Comment:
Can we rename SsoCachedAccessTokenProvider to SsoAccessTokenProvider? The fact that it works on a cache seems like an implementation detail.

Reply:
Since we are going to put the logic of resolving access token back into the inner class of the factory, we won't need this token provider any more.

Reply:
Will do the change.